go/ephemeral/lib.go

```diff
@@ -384,23 +401,55 @@ func (e *EKLib) getOrCreateLatestTeamEKInner(ctx context.Context, teamID keybase
 	// is not expired. It's crucial that this verifies that the latest PTK was
 	// used since we don't want to use a key signed by an old PTK for
 	// encryption.
-	teamEKNeeded, latestGeneration, err := e.newTeamEKNeeded(ctx, teamID, merkleRoot)
+	teamEKNeeded, backgroundGenPossible, latestGeneration, err := e.newTeamEKNeeded(ctx, teamID, merkleRoot)
 	if err != nil {
 		return teamEK, err
 	} else if teamEKNeeded {
-		publishedMetadata, err := publishNewTeamEK(ctx, e.G(), teamID, merkleRoot)
-		if err != nil {
-			return teamEK, err
+		// Our current teamEK is expired but otherwise valid, so let's create
+		// the new teamEK in the background.  For large teams and an empty UPAK
+		// cache it can be expensive to do this calculation and it's
+		// unfortunate to block message sending while we otherwise have access
+		// to a working teamEK.
+		if backgroundGenPossible {
+			go func() {
+				if e.backgroundCreationTestCh != nil {
+					select {
+					case <-e.backgroundCreationTestCh:
+					}
+				}
+
+				publishedMetadata, err := publishNewTeamEK(ctx, e.G(), teamID, merkleRoot)
+				// Grab the lock once we finish publishing so we do don't block
+				e.Lock()
+				defer e.Unlock()
```

Comment:
I think there's a race here. Thread 1 publishes a new teamEK and gets to this line. Thread 2 then publishes _yet another_ teamEK (at the following generation) and gets to this line. If thread 2 happens to grab this lock first, then thread 1 will grab it after and overwrite thread 2's key.

It's a really unlikely race, but maybe we want a map of per-team locks?

Reply:
i don't think it's possible since only one thread can be in here at a time since the `CreationInProgress` flag is set, blocking thread2 from reaching the code to allow background generation